File: src/main/java/dragonbook/inter/stmt/SetElem.java
package dragonbook.inter.stmt;

import dragonbook.error.ParseError;
import dragonbook.inter.expr.Access;
import dragonbook.inter.expr.Expr;
import dragonbook.inter.expr.Id;
import dragonbook.symbols.Array;
import dragonbook.symbols.Type;

public record SetElem(Access access, Expr expr) implements Stmt {

    public SetElem {
        checkTypes(access.type(), expr.type());
    }

    private void checkTypes(Type type1, Type type2) {
        if (type1 instanceof Array || type2 instanceof Array) throw new ParseError("type error");
        else if (type1 == type2) return;
        else if (type1.isNumeric() && type2.isNumeric()) return;
        throw new ParseError("type error");
    }

    public Id array() {
        return access.array();
    }

    public Expr index() {
        return access.index();
    }
}

File: src/main/kotlin/dragonbook/inter/stmt/SetElem.kt
package dragonbook.inter.stmt

import dragonbook.error.ParseError
import dragonbook.inter.expr.Access
import dragonbook.inter.expr.Expr
import dragonbook.inter.expr.Id
import dragonbook.symbols.Array

data class SetElem(val access: Access, val expr: Expr) : Stmt {

    init {
        validate()
    }

    private fun validate() {
        if (access.type is Array || expr.type is Array) throw ParseError(node = this)
        else if (access.type == expr.type) expr.type
        else if (access.type.isNumeric() && expr.type.isNumeric()) expr.type
        else throw ParseError(node = this)
    }

    val array: Id get() = access.array
    val index: Expr get() = access.index

    override fun toString(): String = "$array[$index] = $expr"
}